packages/lesswrong/components/ultraFeed/UltraFeedPostDialog.tsx

```diff
@@ -0,0 +1,233 @@
+    fontSize: '1.4rem',
+    fontWeight: 600,
+    opacity: 0.8,
+    lineHeight: 1.15,
+    textWrap: 'balance',
+    width: '100%',
+    '&:hover': {
+      opacity: 0.9,
+      textDecoration: 'none',
+    },
+    [theme.breakpoints.down('sm')]: {
+      fontSize: '1.6rem',
+    },
+  },
+  closeButton: {
+    fontFamily: theme.palette.fonts.sansSerifStack,
+    color: theme.palette.grey[500],
+    fontWeight: 600,
+    fontSize: "1.1rem",
+    cursor: 'pointer',
+    padding: 8,
+    '&:hover': {
+      color: theme.palette.grey[700],
+    },
+  },
+  dialogPaper: {
+    maxWidth: 765,
+    margin: 4
+  },
+  loadingContainer: {
+    display: "flex",
+    justifyContent: "center",
+    padding: "40px 0",
+  },
+  voteContainer: {
+    display: 'flex',
+    alignItems: 'center',
+    justifyContent: 'center',
+    padding: 16,
+    fontFamily: `${theme.palette.fonts.sansSerifStack} !important`,
+  },
+  voteBottom: {
+    display: 'flex',
+    justifyContent: 'center',
+    position: 'relative',
+    fontSize: 42,
+    textAlign: 'center',
+    marginBottom: 40,
+    "@media print": { display: "none" },
+    '& h1': {
+      fontFamily: `${theme.palette.fonts.sansSerifStack} !important`,
+    }
+  },
+}));
+
+type UltraFeedPostDialogProps = {
+  postId?: string;
+  post?: never;
+  onClose: () => void;
+  textFragment?: string;
+} | {
+  postId?: never;
+  post: PostsPage | UltraFeedPostFragment;
+  onClose: () => void;
+  textFragment?: string;
+}
+
+const UltraFeedPostDialog = ({
+  postId,
+  post,
+  onClose,
+  textFragment,
+}: UltraFeedPostDialogProps) => {
+  const { LWDialog, FeedContentBody, Loading, CommentsListSection, PostsVote } = Components;
+  const classes = useStyles(styles);
+
+  const { document: fetchedPost, loading: loadingPost } = useSingle({
+    documentId: postId,
+    collectionName: "Posts",
+    fragmentName: "UltraFeedPostFragment",
+    skip: !!post, 
+  });
+
+  const { results: comments, loading: isCommentsLoading, totalCount: commentsTotalCount } = useMulti({
+    terms: {
+      view: "postCommentsTop",
+      postId: postId ?? post?._id,
+      limit: 100, // TODO: add load more
+    },
+    collectionName: "Comments",
+    fragmentName: "CommentsList",
+    skip: !(postId ?? post?._id),
+    enableTotal: true,
+  });
+
+  const fullPost = post ?? fetchedPost;
+  const isLoading = !!postId && loadingPost && !post; 
+
+  const previousHashRef = useRef<string | null>(null);
+  const appliedHashRef = useRef<string | null>(null);
+
+  useEffect(() => {
+    let isMounted = true; 
+    if (!isLoading && textFragment && textFragment !== appliedHashRef.current) {
+      const timeoutId = setTimeout(() => {
+        if (isMounted && !isLoading && textFragment && textFragment !== appliedHashRef.current) {
+            if (previousHashRef.current === null) { 
+               previousHashRef.current = window.location.hash; 
+            }
+            
+            window.location.hash = textFragment; 
```

Comment:
this should almost certainly be a `navigate`

Reply:
no, navigate doesn't work here